### ISSUE TITLE
Sprintf always evaluated no matter the log level

### ISSUE THREAD

**Issue by gunnihinn** [jcelliott/lumber#9]
The way the various level functions (`Info`, `Debug`, `Warn` etc) contain this code:

    fmt.Sprintf(format, v...)

This means that the `Sprintf` call is always evaluated, no matter what the log level is. This can be very problematic when the package is used with expensive data structures and we want detailed debug messages that should be ignored the rest of the time.

For example, I was using your package with a linter I'm building. This linter traverses a tree and I had seeded it with `logger.Debug` calls that printed out the whole tree recursively. When I changed the underlying parser to output a bigger tree, my program all of the sudden ran on the order of 10 minutes instead of a couple of seconds. The cause was the calls to `logger.Debug` which were evaluating the string printouts of the entire tree about once per every vertex in the tree, even when the log level was set to less than `DEBUG`.

If users have to worry that writing debug messages will affect the performance of their program they won't do it. That defeats the purpose of multilevel logging.

A possible solution is to order the log levels and wrap the logging calls inside the level functions inside a function that checks the current log level.

**Comment by gunnihinn** [comment]
Hum... wait a minute, I'm the idiot here, that expensive function was already called every time I called the `Debug` function to pass a string to it as an argument.

Nevermind me.

**Comment by jcelliott** [comment]
You are correct about functions being called before their results are passed into the logging functions. The way to work with expensive debuging function calls is by using the `Is*` functions. For example:

```
if lumber.IsDebug() {
    expensiveDebuggingCall()
}
```

You were also correct about the printf arguments being evaluated when they shouldn't have been. I fixed that a while back for the filelogger, bug forgot to change it for the consolelogger, so I have just updated that.
